### DISCUSSION TITLE
Always on top window

### DISCUSSION THREAD

**Discussion by tux7k** [fltk-rs/fltk-rs#826] [Q&A]
Is it possible to make a window always on top? I've thought about using the Windows API but I couldn't find a way to do it.

**Comment by MoAlyousef** [comment]
Yeah, you can use `wind.make_modal(true)` before showing the window.

**Reply by tux7k** [reply]
Trying so doesn't work, if it helps I'm trying to get it to be on top of all windows

**Reply by woj-tek** [reply]
Tried that on macOS and it didn't work. Any other options?

Java has simple [`setAlwaysOnTop()`](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/Window.html#setAlwaysOnTop(boolean)) and it would be awesome to something similar in fltk :-)

**Reply by MoAlyousef** [reply]
You can try win.set_on_top()

**Reply by woj-tek** [reply]
Hmm... I was looking for something in https://docs.rs/fltk/latest/fltk/window/struct.DoubleWindow.html (which `Window::default()` defaults to and I was following basic examples) but it was in https://docs.rs/fltk/latest/fltk/window/struct.SingleWindow.html#method.set_on_top

If I try to swap default for `SingleWindow`:

rust:
```
…
    let app = app::App::default();

    let mut window = SingleWindow::new(100, 100, 400, 300, "Hello from rust");
    window.set_on_top();
…
```

it compiles but throws:

```
assertion failed: !ptr.is_null()
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: <fltk::window::SingleWindow as fltk::prelude::WindowExt>::raw_handle
             at /Users/wojtek/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fltk-1.4.18/src/macros/window.rs:278:25
   4: fltk::window::SingleWindow::set_on_top
             at /Users/wojtek/.cargo/registry/src/index.crates.io-6f17d22bba15001f/fltk-1.4.18/src/window.rs:264:32
```

and my extremely rudimentary rust knowledge goes out of the window at this point :-)

**Reply by MoAlyousef** [reply]
It requires that it be called after win.show(). Since this calls some native cocoa function, only when the window is shown will the platform window be valid to the OS.

**Reply by woj-tek** [reply]
Oh, that works (how did you figure it out? It's not obvious from documentation).

Thanks!
Now I only have to adjust code to SingleWindow API :-)

**Reply by MoAlyousef** [reply]
I’ve run into this some time ago with FLTK :)
Not that set_on_top only works for macos.

**Comment by MoAlyousef** [comment]
Depends on the operating system and window manager. For windows you can use SetWindowPos:
```rust
fn make_topmost(w: &window::Window) {
    assert!(w.shown());
    use std::os::raw::*;
    const SWP_NOSIZE: u32 = 0x0001; // ignore the width and height params
    const SWP_NOMOVE: u32 = 0x0002; // ignore the x and y params
    pub type HWND = *mut c_void;
    extern "C" {
        pub fn SetWindowPos(
            hwnd: HWND,
            hwnd: HWND,
            x: i32,
            y: i32,
            cx: i32,
            cy: i32,
            uflags: u32,
        ) -> i32;
    }
    unsafe {
        let handle = w.raw_handle();
        let val: *mut i64 = std::mem::transmute(-1 as i64);
        SetWindowPos(handle as _, val as _, 0, 0, 0, 0, SWP_NOMOVE | SWP_NOSIZE);
    }
}
```

**Reply by tux7k** [reply]
Ok I will try ASAP
